File: flink-sql/src/main/scala/com/otis/flinksql/quickStart/Demo1.scala
package com.otis.flinksql.quickStart

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala._
import org.apache.flink.table.api.Table
import org.apache.flink.table.api.scala._

object Demo1 {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    env.setParallelism(1)

    val stream: DataStream[SensorReading] = env.addSource(new SensorSource)

    //1. 创建表环境
    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)

    //2. 基于tableEnv,将流转换成表
    val table: Table = tableEnv.fromDataStream(stream)

    /**
     * 3.1 这样就可以调用Table（java的接口）接口的方法，table api转换操作
     */

    val tableResult1: Table = table.select("id,temperature")
      .filter("id == 'sensor_1'")

    /**
     * 3.2 直接写sql实现
     */
    //首先注册一张表
    val myTable = tableEnv.registerTable("tableName", table)
    //sql他复杂太长，可以用三引号来自动换行
    val tableResult2: Table = tableEnv.sqlQuery(
      """
        |select id,temperature
        |from tableName
        |where id='sensor_1'
        |""".stripMargin)

    val unit: DataStream[(String, Double)] = tableResult2.toAppendStream[(String, Double)]
    unit.print()

    env.execute("sql demo")
  }
}

File: flink-sql/src/main/java/com/qinghua/java/quickStart/Demo1.java
package com.qinghua.java.quickStart;

import com.otis.flinksql.quickStart.SensorReading;
import com.otis.flinksql.quickStart.SensorSource;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.java.StreamTableEnvironment;

public class Demo1 {
    public static void main(String[] args) {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        DataStreamSource<SensorReading> stream = env.addSource(new SensorSource());
        //创建表环境
        StreamTableEnvironment tableEnvironment = StreamTableEnvironment.create(env);

//        tableEnvironment.sqlQuery()

        //把流转换成表
        Table table = tableEnvironment.fromDataStream(stream);

        Table resultTable = table.select("id,temperature")
                .filter("id == 'sensor_1'");

    }
}
